addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlMessageDeserializer.java

```diff
@@ -0,0 +1,106 @@
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+
+/**
+ * Class {@link NikoHomeControlMessageDeserializer} deserializes all json messages from Niko Home Control. Various json
+ * message
+ * formats are supported. The format is selected based on the content of the cmd and event json objects.
+ *
+ * @author Mark Herwege
+ *
+ */
+class NikoHomeControlMessageDeserializer implements JsonDeserializer<NhcMessageBase> {
+
+    @Override
+    public NhcMessageBase deserialize(final JsonElement json, final Type typeOfT,
+            final JsonDeserializationContext context) throws JsonParseException {
+
+        final JsonObject jsonObject = json.getAsJsonObject();
+
+        try {
+            String cmd = null;
+            String event = null;
+            if (jsonObject.has("cmd")) {
+                cmd = jsonObject.get("cmd").getAsString();
+            }
+            if (jsonObject.has("event")) {
+                event = jsonObject.get("event").getAsString();
+            }
+
+            JsonElement jsonData = null;
+            if (jsonObject.has("data")) {
+                jsonData = jsonObject.get("data");
+            }
+
+            NhcMessageBase message = null;
+
+            if (jsonData != null) {
+
+                if (jsonData.isJsonObject()) {
+
+                    message = new NhcMessageMap();
+
+                    Map<String, String> data = new HashMap<>();
+                    for (Entry<String, JsonElement> entry : jsonData.getAsJsonObject().entrySet()) {
+                        data.put(entry.getKey(), entry.getValue().getAsString());
+                    }
+                    ((NhcMessageMap) message).setData(data);
+
+                } else if (jsonData.isJsonArray()) {
+
+                    JsonArray jsonDataArray = jsonData.getAsJsonArray();
+
+                    message = new NhcMessageListMap();
+
+                    List<HashMap<String, String>> dataList = new ArrayList<>();
+                    for (int i = 0; i < jsonDataArray.size(); i++) {
+                        JsonObject jsonDataObject = jsonDataArray.get(i).getAsJsonObject();
+
+                        Map<String, String> data = new HashMap<>();
+                        for (Entry<String, JsonElement> entry : jsonDataObject.entrySet()) {
+                            data.put(entry.getKey(), entry.getValue().getAsString());
+                        }
+                        dataList.add((HashMap<String, String>) data);
+                    }
+                    ((NhcMessageListMap) message).setData(dataList);
+                }
+
+            }
+
+            if (message != null) {
+                message.setCmd(cmd);
+                message.setEvent(event);
+            } else {
+                throw new JsonParseException("Unexpected Json type");
+            }
+
+            return message;
+
+        } catch (IllegalStateException | ClassCastException e) {
```

Comment:
I don't want the binding to completely fail when what I get back from the controller is different from what I expect. Debug logging will show me exactly what came back and will tell me if it cannot be interpreted. IllegalStateException can be thrown from the nested getAsJsonObject() calls when the inner element would be a json array or primitive, rather then an object. Similarly ClassCastException would occur if the getAsString() call is not on a primite, but an object or array.
A number of messages I get from the controller, but do not interpret at the moment, have formats that cannot be deserialized with the current logic yet. The nested logic would have to be refined. Also, I am not 100% sure all versions of the controller software use exactly the same json formats. This also shields from that. The debug log will show an 'Unsupported json type' message.
I learned the structure from this: [http://www.javacreed.com/gson-deserialiser-example/](http://www.javacreed.com/gson-deserialiser-example/). As the nesting can be complex and unpreditable, I only went with what I know for know, and will extend this part if I implement extra functionality.